src/main/java/org/springframework/cli/command/ProjectCommands.java

```diff
@@ -80,62 +82,51 @@ public void projectAdd(
-		if (projectRepositories != null) {
-			rows = projectRepositories.stream()
-				.map(tr -> new String[] { tr.getName(),
-						Objects.requireNonNullElse(tr.getDescription(), ""),
-						tr.getUrl(),
-						"",
-						(Objects.requireNonNullElse(tr.getTags(), "")).toString() }
-			);
-		}
-		else {
-			rows = Stream.empty();
-		}
-		List<String[]> allRows = rows.collect(Collectors.toList());
+		List<ProjectRepository> repos = upCliUserConfig.getProjectRepositories().getProjectRepositories();
+		List<ProjectRepositoryData> projectRepositories = (repos == null ? Stream.<ProjectRepository>empty() : repos.stream())
+				.map(pr -> new ProjectRepositoryData(pr.getName(), pr.getUrl(), pr.getDescription(), pr.getTags(), null))
+				.collect(Collectors.toList());
 
 		// List projects that are contained in catalogs that the user had added using the `project-catalog add` command
 		List<ProjectCatalog> projectCatalogs = upCliUserConfig.getProjectCatalogs().getProjectCatalogs();
 		for (ProjectCatalog projectCatalog : projectCatalogs) {
 			String url = projectCatalog.getUrl();
 			Path path = sourceRepositoryService.retrieveRepositoryContents(url);
 			YamlConfigFile yamlConfigFile = new YamlConfigFile();
-			projectRepositories = yamlConfigFile.read(Paths.get(path.toString(),"project-catalog.yml"),
-					ProjectRepositories.class).getProjectRepositories();
-			Stream<String[]> fromCatalogRows;
-			if (projectRepositories != null) {
-				fromCatalogRows = projectRepositories.stream()
-						.map(tr -> new String[] {
-								tr.getName(),
-								Objects.requireNonNullElse(tr.getDescription(), ""),
-								tr.getUrl(),
-								projectCatalog.getName(),
-								(Objects.requireNonNullElse(tr.getTags(), "")).toString() }
-						);
-			} else {
-				fromCatalogRows = Stream.empty();
+			for (ProjectRepository pr : yamlConfigFile.read(Paths.get(path.toString(),"project-catalog.yml"), ProjectRepositories.class).getProjectRepositories()) {
+				projectRepositories.add(new ProjectRepositoryData(pr.getName(), pr.getUrl(), pr.getDescription(), pr.getTags(), projectCatalog.getName()));
 			}
-			List<String[]> projectCatalogRows = fromCatalogRows.collect(Collectors.toList());
-			allRows.addAll(projectCatalogRows);
-
 			// clean up temp files
 			try {
 				FileSystemUtils.deleteRecursively(path);
 			} catch (IOException ex) {
 				logger.warn("Could not delete path " + path, ex);
 			}
+		}
 
+		if (json) {
+			return objectMapper.writeValueAsString(projectRepositories);
 		}
 
-		String[][] data = Stream.concat(header, allRows.stream()).toArray(String[][]::new);
+		Stream<String[]> header = Stream.<String[]>of(new String[] { "Name", "Description",  "URL", "Catalog", "Tags" });
+
+		// Retrieve project that were registered using the `project add` command and stored locally
+		Stream<String[]> rows = projectRepositories.stream()
+			.map(tr -> new String[] { tr.name(),
+					Objects.requireNonNullElse(tr.description, ""),
+					tr.url(),
+					"",
```

Comment:
If this is left empty then `Catalog` column in table don't have content anymore. It used to come from `projectCatalog.getName()` in a code you refactored.

Reply:
Yes, i missed that - good catch. Should be fixed now.